content/unroll-default-arguments.md

```diff
@@ -0,0 +1,933 @@
+      args: Seq[String],
+      allowPositional: Boolean,
+      allowRepeats: Boolean,
+      totalWidth: Int,
+      printHelpOnExit: Boolean,
+      docsOnNewLine: Boolean,
+      autoPrintHelpAndExit: Option[(Int, PrintStream)],
+      customName: String,
+      customDoc: String,
+      nameMapper: String => Option[String]
+  ): Either[String, T] = constructEither(
+    args,
+    allowPositional,
+    allowRepeats,
+    totalWidth,
+    printHelpOnExit,
+    docsOnNewLine,
+    autoPrintHelpAndExit,
+    customName,
+    customDoc,
+    sorted = true,
+    nameMapper = nameMapper
+  )
+```
+
+Apart from being extremely verbose and full of boilerplate, like any boilerplate this is
+also extremely error-prone. Bugs like [com-lihaoyi/mainargs#106](https://github.com/com-lihaoyi/mainargs/issues/106)
+slip through when a mistake is made in that boilerplate. These bugs are impossible to catch 
+using a normal test suite, as they only appear in the presence of version skew. The above code 
+snippet actually _does_ have such a bug, that the test suite _did not_ catch. See if you can 
+spot it! 
+
+Sebastien Doraene's talk [Designing Libraries for Source and Binary Compatibility](https://www.youtube.com/watch?v=2wkEX6MCxJs)
+explores some of the challenges, and discusses the workarounds. 
+
+
+## Requirements
+
+### Backwards Compatibility
+
+Given:
+
+* Two libraries, **Upstream** and **Downstream**, where **Downstream** depends on **Upstream**
+
+* If we use a _newer_ version of **Upstream** which contains an added
+  default parameter together with an _older_ version of **Downstream** compiled
+  against an _older_ version of **Upstream** before that default parameter was added
+
+* The behavior should be binary compatible and semantically indistinguishable from using
+  a verion of **Downstream** compiled against the _newer_ version of **Upstream**
+
+**Note:** we do not aim for _Forwards_ compatibility. Using an _older_ 
+version of **Upstream** with a _newer_ version of **Downstream** compiled against a 
+_newer_ version of **Upstream** is not a use case we want to support. The vast majority
+of OSS software does not promise forwards compatibility, including software such as
+the JVM, so we should just follow suite
+
+### All Overrides Are Equivalent
+
+All versions of an `@unroll`ed method `def foo` should have the same semantics when called 
+with the same parameters. We must be careful to ensure:
+
+1. All our different method overrides point at the same underlying implementation
+2. Abstract methods are properly implemented, and no method would fail with an 
+   `AbstractMethodError` when called
+3. We properly forward the necessary argument and default parameter values when
+   calling the respective implementation.
+
+## Proposed solution
+
+
+The proposed solution is to provide a `scala.annotation.unroll` annotation, that
+can be applied to methods `def`s, `class` constructors, or `case class`es to generate
```

Comment:
In the description I've seen only `@unroll` in case class constructor.

Would it be possible to `@unroll` a case class as such? Meaning it would `@unroll` from the first parameter with default value.

Reply:
@sideeffffect can you elaborate on what you mean by "*`@unroll` a whole case class*" and how it differs from the behavior described under **Case Classes**?

Reply:
I guess this is the idea:
```scala
@unroll case class Data(x: Int, y: Int = 0, z: Int = 0)
```
So it is unrolled from y to z.

Reply:
In the current proposal that would be expressed

```scala
case class Data(x: Int, @unrollAll y: Int = 0, z: Int = 0)
```

Reply:
Yes, but I guess the idea is that it may be easier to spot/read if it's annotation on the whole class. I'm not sure.

Reply:
Also, I'm still a bit hesitant if the term "unroll" is the best possible, but I have failed to come up with something better. It would be nice if the term said something concrete about what is happening, like "autoOverloaded" or something.

It is not that the name matters very much fpr lib authors, but for newcomers that come across "cryptic" annotations, it can be difficult.

Reply:
If I duckduck "unroll" I'm suggested a service to "unsubscribe from emails, instantly" and much other stuff unrelated to programming.

Reply:
Perhaps `@compatOverload` can reveal both something about the intent (compatibility) and mechanism (overloading). (I think "unroll" says little about neither intent nor mechanism - that's why I'm brainstorming names...)

Reply:
And if the annotation is allowed outside the thing it would cover the typical use case without cluttering the paramlist, so the reading and grokking of the thing is not disturbed:
```scala
@compatOverload 
case class Thing(x: Int, y: Int = 0, z: Int = 0)

@compatOverload 
def thing(x: Int, y: Int = 0, z: Int = 0)
```

Reply:
Indeed, that's exactly what I had in mind, that
```scala
@unroll case class Data(x: Int, y: Int = 0, z: Int = 0)
```
would behave as
```scala
case class Data(x: Int, @unrollAll y: Int = 0, z: Int = 0)
```
does currently.

I think many people would prefer this over thinking where exactly in the constructor list to put the `@unrollAll`. I certainly would.

Reply:
Main view here is that "*unroll everything*" is most often *not* what you want. 

You only want to `@unroll` the individual parameters that correspond to a new release of the library, which is a usually subset of the default parameters a method or case class. Often methods are defined with default parameters to begin with, and new versions of a library may add more than one default parameter to a method. You only want to `@unroll` the first default parameter for each new version of the library.

It doesn't do any harm to "`@unroll` everything", but it is wasteful in terms of generated code, and thus both compile times and classfile size. This was discussed when the SIP was initially proposed, and we decided to go with `@unroll` on specific parameters rather than `@unroll`ing entire `def` or `case class`es as the default as it better reflects what you would actually want in reality despite being a bit more verbose and intrusive

Reply:
Could we then have both the shorter `@unroll` on individual parameters and a more explanatory similar to `@unrollDefaultParams` on the whole thing to cover both use cases?

You say "most often" but I'd assume that if you plan for binary compatibility then as soon as you add things all the params from the first update would need overloaded forwarders, so it seems to me that the use case of "from now on generate forwarders for all added params" would be the most common?

Reply:
> You say "most often" but I'd assume that if you plan for binary compatibility then as soon as you add things all the params from the first update would need overloaded forwarders, so it seems to me that the use case of "from now on generate forwarders for all added params" would be the most common?

That is only true if you add one default parameter at a time per stable release of your library. That is sometimes the case, sometimes isn't. Even if no single PR adds multiple default parameters, a stable release often includes multiple PRs. If a stable release includes multiple default parameters, then you should only generate an overload for the first one and not all of them

Reply:
Another thing to consider is that `@unrollAll` *still has to specify a parameter to start unrolling from*. Often methods come with default parameters right out of the box, and you do *not* want to unroll any of those. You only want to unroll default parameters added in *subsequent stable releases*, which means you need to specify a parameter from which unrolling will take place, which means putting the annotation on the entire method `def`/`class`/`case class` does not allow us to convey enough information to make that work